protocols/gossipsub/src/layer.rs

```diff
@@ -0,0 +1,1788 @@
+    peer_topics: HashMap<PeerId, (SmallVec<[TopicHash; 16]>, NodeType)>,
+
+    /// Overlay network of connected peers - Maps topics to connected gossipsub peers.
+    mesh: HashMap<TopicHash, Vec<PeerId>>,
+
+    /// Map of topics to list of peers that we publish to, but don't subscribe to.
+    fanout: HashMap<TopicHash, Vec<PeerId>>,
+
+    /// The last publish time for fanout topics.
+    fanout_last_pub: HashMap<TopicHash, Instant>,
+
+    /// Message cache for the last few heartbeats.
+    mcache: MessageCache,
+
+    // We keep track of the messages we received (in the format `string(source ID, seq_no)`) so that
+    // we don't dispatch the same message twice if we receive it twice on the network.
+    received: CuckooFilter<DefaultHasher>,
+
+    /// Heartbeat interval stream.
+    heartbeat: Interval,
+
+    /// Marker to pin the generics.
+    marker: PhantomData<TSubstream>,
+}
+
+impl<TSubstream> Gossipsub<TSubstream> {
+    /// Creates a `Gossipsub` struct given a set of parameters specified by `gs_config`.
+    pub fn new(local_peer_id: PeerId, gs_config: GossipsubConfig) -> Self {
+        Gossipsub {
+            config: gs_config.clone(),
+            events: VecDeque::new(),
+            local_peer_id,
+            topic_peers: HashMap::new(),
+            peer_topics: HashMap::new(),
+            mesh: HashMap::new(),
+            fanout: HashMap::new(),
+            fanout_last_pub: HashMap::new(),
+            mcache: MessageCache::new(gs_config.history_gossip, gs_config.history_length),
+            received: CuckooFilter::new(),
+            heartbeat: Interval::new(
+                Instant::now() + gs_config.heartbeat_initial_delay,
+                gs_config.heartbeat_interval,
+            ),
+            marker: PhantomData,
+        }
+    }
+
+    /// Subscribes to a topic.
+    ///
+    /// Returns true if the subscription worked. Returns false if we were already subscribed.
+    pub fn subscribe(&mut self, topic: Topic) -> bool {
+        debug!("Subscribing to topic: {:?}", topic);
+        if self.mesh.get(&topic.hash()).is_some() {
+            debug!("Topic: {:?} is already in the mesh.", topic);
+            return false;
+        }
+
+        // send subscription request to all floodsub and gossipsub peers
+        for (gossip_peers, flood_peers) in self.topic_peers.get(&topic.hash()) {
+            for peer in flood_peers.iter().chain(gossip_peers) {
+                debug!("Sending SUBSCRIBE to peer: {:?}", peer);
+                self.events.push_back(NetworkBehaviourAction::SendEvent {
+                    peer_id: peer.clone(),
+                    event: GossipsubRpc {
+                        messages: Vec::new(),
+                        subscriptions: vec![GossipsubSubscription {
+                            topic_hash: topic.hash().clone(),
+                            action: GossipsubSubscriptionAction::Subscribe,
+                        }],
```

Comment:
Yes, I have been thinking a similar thing (grouping messages into a single `GossipsubRpc`). There were two strategies that came to mind:
1) For some specific subscriptions/messages/control messages, we don't immediately add them to `Gossipsub.events` instead send them during the heartbeat (which occurs every second by default). Giving us the chance to accumulate various messages and subscriptions before sending them across the wire. The messages would have to be carefully chosen as we want to minimize latency for messages propagating across the network, but perhaps control messages could be delayed and accumulated. 
2) We could check the `send.events` list and add onto any other GossipsubRpc events being sent to the same peer. I imagine this may be a rare find as the poll consumes the events quite rapidly. 
The concept of control message `piggybacking` is related here. 

Is your point about the `Arc` related more related minimizing the memory overhead of creating the structs?

Reply:
> Is your point about the `Arc` related more related minimizing the memory overhead of creating the structs?

Yes.

Reply:
I've added an Arc  for all events. This required mapping the `NetworkBehaviour<Arc<_>, _>` to `NetworkBehaviour<_,_>` in the `NetworkBahviour` `poll()`. 
An alternate solution was to implement an `ArcGossipsubRpc { inner: Arc<GossipsubRpc> }` and implement `OutboundUpgrade` on this new struct and use this for the events. This struct would be cloning the inner Arc's when sending across the network, so decided to localize the code changes by perform the cloning of the underlying RPC structs in the `poll()` function. 
See https://github.com/libp2p/rust-libp2p/pull/898/commits/ed523b8671122eac8df4e30e0fa9e6bb9428dca0